### PR TITLE
xcode: die unless Xcode/CLT up-to-date on prerelease macOS

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Not only have several maintainers suggested this is a good idea, including myself repeatedly, but at least a handful of users have said they wouldn't hate the idea so let's follow up better late than never.

Obvious test is obvious:
screen shot 2016-08-30 at 04 02 25

### CODE REVIEW

**Comment by DomT4** [comment]
Huh. That failure did not happen locally.

**Comment by DomT4** [comment]
Oh, I see why. Can I wrap it in a simple `if OS.mac?` or are people going to insist I write a separate macOS compatibility layer for the `install` command? 🙈.

File: Library/Homebrew/cmd/install.rb [hunk 1]
```diff
@@ -227,6 +227,12 @@ def check_development_tools
       out = checks.send(check)
       opoo out unless out.nil?
     end
+    if OS.mac? && MacOS.prerelease?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: Pull this logic into `all_development_tools_checks` instead to avoid duplication and using `OS.mac?` here.

**Comment by MikeMcQuaid** [comment]
One comment but otherwise agree with it.

> or are people going to insist I write a separate macOS compatibility layer for the install command? 🙈.

Usually yes but in this case you can just pull into diagnostic instead; I don't think we need a separate set of "strict" checks.

**Comment by DomT4** [comment]
> Pull this logic into all_development_tools_checks instead to avoid duplication and using OS.mac? here.

Can you clarify this a bit?

I didn't lump the Xcode/CLT checks into `all_development_tools_checks` because `all_development_tools_checks` also includes a check as to whether you're running an unsupported version of macOS, so making that section a hard block would stop people running `install` on 10.5-10.8 and 10.12.

Either that or we throw the blob into `all_development_tools_checks` but don't make it fatal, retaining the `oppo` and hope people actually notice it/take it seriously.

I presume I'm misunderstanding how simple you want this, so thought I'd just ask 😅.

**Comment by MikeMcQuaid** [comment]
I missed the `odie` difference 👍 

Just want to avoid have `if OS::Mac.prerelease?` or `if MacOS.` anywhere outside `Library/Homebrew/extend/os/mac` now. It would be good to instead move as much of this logic as possible to a method in `Library/Homebrew/extend/os/mac/diagnostic.rb` which could also do the `odie` and have a no-op in `diagnostic.rb`. e.g. have a `Diagnostic::Checks. check_development_tools` method that's a no-op for non-OS X or consider if this could/should live in `DevelopmentTools` instead.

**Pull request merged**